tests/test_batch_executor.py

```diff
@@ -0,0 +1,121 @@
+import random
+from concurrent.futures import Future
+from typing import Any, Generator, List, Tuple
+
+import pytest
+
+from vcap.batch_executor import BatchExecutor, _Request
+
+
+@pytest.fixture()
+def batch_executor():
+    """To use this fixture, replace batch_executor.batch_fn with your own
+    batch function."""
+
+    def batch_fn(inputs):
+        raise NotImplemented
+
+    batch_executor = BatchExecutor(batch_fn=batch_fn)
+    yield batch_executor
+    batch_executor.close()
+
+
+def batch_fn_base(inputs: List[int], raises: bool) \
+        -> Generator[Any, None, None]:
+    """Process results and yield them as they are processed
+    :param inputs: A list of inputs
+    :param raises: If True, raises an error on the 5th input
+    """
+    for i in inputs:
+        if i == 5 and raises:
+            raise RuntimeError("Oh no, a batch_fn error has occurred!")
+        yield i * 100
+
+
+def batch_fn_returns_generator(inputs: List[int]) \
+        -> Generator[Any, None, None]:
+    return (o for o in batch_fn_base(inputs, raises=False))
+
+
+def batch_fn_returns_generator_raises(inputs: List[int]) \
+        -> Generator[Any, None, None]:
+    return (o for o in batch_fn_base(inputs, raises=True))
+
+
+def batch_fn_returns_list(inputs: List[int]) -> List[Any]:
+    """Process results and yield them at the end, as a list."""
+    return list(batch_fn_base(inputs, raises=False))
+
+
+def batch_fn_returns_list_raises(inputs: List[int]) -> List[Any]:
+    return list(batch_fn_base(inputs, raises=True))
+
+
+@pytest.mark.parametrize(
+    argnames=["batch_fn", "expect_partial_results"],
+    argvalues=[
+        (batch_fn_returns_generator_raises, True),
+        (batch_fn_returns_list_raises, False)
+    ]
+)
+def test_exceptions_during_batch_fn(
+        batch_executor, batch_fn, expect_partial_results):
+    """Test that BatchExecutor catches exceptions that occur in the batch_fn
+    and propagates them through the requests Future objects.
+
+     If an exception occurs after processing some of the batch, the expectation
```

Comment:
🚨 IMPORTANT 🚨 Indentation is one space too far to the right!!!

Reply:
FIXED FIXED ASAP ALL CLEAR!